include/ginkgo/core/base/lin_op.hpp

```diff
@@ -753,7 +753,7 @@ public:                                                                \
  *     GKO_ENABLE_LIN_OP_FACTORY(MyLinOp, my_parameters, Factory) {
  *         // a factory parameter named "my_value", of type int and default
  *         // value of 5
- *         int GKO_FACTORY_PARAMETER(my_value, 5);
+ *         int GKO_FACTORY_PARAMETER_SCALAR(my_value, 5);
```

Comment:
Maybe add an example with `GKO_FACTORY_PARAMETER_VECTOR` to also showcase this one?

Reply:
Good point!

Reply:
I am confused with `std::pair` example.
I feel std::pair is a scalar containing two elements like complex type.
maybe like `std::pair<int, int> GKO_FACTORY_PARAMETER_SCALAR(my_value, make_pair(5, 5))`?

Reply:
The difference between _SCALAR and _VECTOR comes from how cudafe++ translates the files. I'm not entirely sure about the formal criteria, but for `int` and other primitive types, the parameter pack expansion gets translated incorrectly (Into a C-style cast instead of a constructor call), while for composite types like `std::pair, std::vector, ...` it correctly invokes the constructor. The semantics of this are a bit fuzzy, though, that is true. You could also use SCALAR, but then you would have to use make_pair or braces { ... } for the default argument or with_my_value.